### ISSUE TITLE
[RFC] Document user-provided compile-time flags

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#11383]
This is a proposal about documenting user-provided compile-time flags in the standard library and compiler.

We're planning to introduce feature flags more frequently in the future, as an opt-in mechanism for breaking changes and previews before they become enabled by default. So the number of flags is expected to change relatively often.

We need to document which compile-time flags are exposed by the standard library and compiler. There is no good opportunity to do that directly in the source code because flags are never explicitly declared, they're just queried via `flag?` macro calls all over the code base. Extracting their documentation in a central location makes sense. This should be inside the repository (as opposed to https://github.com/crystal-lang/crystal-book fo example), to change the documentation in sync with changes to the flags.

A CSV file seems like a decent solution because it's a format for tabular data. It's easily editable, machine readable and there's a nice rendering on Github. YAML would be an alternative which allows more flexibility in terms of data format.
For this demonstration I've used a rendered (CSV) table of the currently available user-defined flags:

| flag | component | description | available_since
|-|-|-|-|
| debug_raise | stdlib | Debugging flag for `raise` logic. Prints the backtrace before raising. |
| gc_none | stdlib | Disables garbage collection ([#5314](https://github.com/crystal-lang/crystal/pull/5314)). |
| preview_mt | stdlib | Enables multithreading preview ([#7546](https://github.com/crystal-lang/crystal/pull/7546)). | 0.28.0
| skip_crystal_compiler_rt | stdlib | Exclude Crystal's native `compiler-rt` implementation. |
|i_know_what_im_doing | compiler | Safety guard against involuntarily building the compiler. |
| without_openssl | compiler | Build the compiler without OpenSSL support. |
| without_zlib | compiler | Build the compiler without Zlib support. |
| without_playground | compiler | Build the compiler without playground (`crystal play`). |

See https://github.com/crystal-lang/crystal-book/pull/554 for a related effort to improve documentation of compiler-provided flags. 

Once this PR is merged, the book can reference the list in this repo for available user-provided flags, or even parse it in order to display the contents in a polished format.
The data can also be used in other places, for example the flags could be exposed via the compiler's `--help` command or in the manpages.

The data format is rather slim right now. Most of the currently used flags are static and not expected to change much. `preview_mt` is the only one that is subject to change as a preview flag, but there's no time frame for enhancements yet.

With the arrival of more temporal feature flags, we should probably add more information, such as the expected end of availability.

There's probably also demand for more in-depth documentation of these flags (currently there's just a short description). I'm not sure if that should be included directly in this list (then probably in YAML format), or in a different place, such as individual text files in this repository or the book.

**Comment by straight-shoota** [comment]
I think the compiler build flags and `skip_crystal_compiler_rt` are useful for package maintainers. `gc_none` can be relevant for debugging. IMO only `debug_raise` and `i_know_what_im_doing' are really specific to compiler/stdlib development.

But I'm also convinced that even internal flags deserve to be documented for compiler developers. It's like code comments.

Maybe we could consider adding a property to define the flag audience or something like that. Then we could exclude compiler development flags from public documentation for example. But I think this would be a later enhancement.

**Comment by rdp** [comment]
Somewhat unrelated, is there a place runtime flags like env: CRYSTAL_LOAD_DWARF are documentated, as well?

**Comment by straight-shoota** [comment]
That's a good point, and very closely related. There's currently no general documentation for that as well. So we should probably include environment variables into our considerations. In contrast to feature flags, they probably more stable.

**Comment by HertzDevil** [comment]
I don't think we really have to be able to autogenerate the table of compile-time flags for the reference manual. But there is a completely different use case here: shell auto-completion scripts could use the CSV to list all predefined flags if one presses the Tab key after `-D`. (For that matter, the auto-completion scripts we have are currently very outdated.)
